tests/python/pants_test/backend/python/tasks/test_ctypes_integration.py

```diff
@@ -42,16 +43,47 @@ def test_ctypes_run(self):
 
     self.assertEqual('x=3, f(x)=17\n', pants_run.stdout_data)
 
-  def test_ctypes_binary(self):
+  def test_ctypes_binary_creation(self):
+    """Create a python_binary() with all native toolchain variants, and test the result."""
+    # TODO: this pattern could be made more ergonomic for `enum()`, along with exhaustiveness
+    # checking.
+    for variant in ToolchainVariant.allowed_values:
+      self._assert_ctypes_binary(variant)
+
+  _compiler_names_for_variant = {
+    'gnu': ['gcc', 'g++'],
+    'llvm': ['clang', 'clang++'],
+  }
+
+  # All of our toolchains currently use the C++ compiler's filename as argv[0] for the linker.
+  _linker_names_for_variant = {
+    'gnu': ['g++'],
+    'llvm': ['clang++'],
+  }
+
+  def _assert_ctypes_binary(self, toolchain_variant):
     with temporary_dir() as tmp_dir:
       pants_run = self.run_pants(command=['binary', self._binary_target], config={
         GLOBAL_SCOPE_CONFIG_SECTION: {
           'pants_distdir': tmp_dir,
-        }
+        },
+        'native-build-settings': {
+          'toolchain_variant': toolchain_variant,
+        },
       })
 
       self.assert_success(pants_run)
 
+      # Check that we have selected the appropriate compilers for our selected toolchain variant,
+      # for both C and C++ compilation.
+      for compiler_name in self._compiler_names_for_variant[toolchain_variant]:
+        self.assertIn("selected compiler exe name: '{}'".format(compiler_name),
```

Comment:
I am definitely not a fan of scraping logs to implement tests. This was cheap since there was already an integration test but expensive since it adds production logging of questionable value (it effectively parrots the very well tested option system) and that will surprisingly break tests when altered. Consider ~never doing this when possible.

Reply:
I will make an effort to remember your comment as the concerns you mention are quite bothersome. I would love to implement any alternative -- the info logs here were pretty much exactly for testing purposes and that conflation makes me sick at heart.

I suppose the point of an integration test is not to try to inspect internal state but to examine some properties of the result of a pants run. I was wondering whether it was possible to tag an object file or shared library with metadata that includes e.g. which compiler produced it. The rationale for testing both compilers here is to avoid introducing a breakage in one toolchain, and just changing the option value doesn't seem like it does enough to verify that we are choosing the right compiler.

It's probably possible to propagate this metadata (which compiler/linker was used) into the produced `python_dist()` (by making some edits to BuildLocalPythonDistributions -- not by tagging any native code), which we could then inspect in this test. I will look up whether this is feasible -- adding a "compiler" tag (or something) to a dist seems like a very interesting and concise way to begin the foray into classifying compatibility between e.g. native code cache artifacts. I'm also seeing the implementation of this requiring a relatively small amount of code, which would be feasible to add in this PR or another. **I'm working on this approach now to avoid the addition of these info-level logs.** I'm thinking of potentially storing another file in `data_files` which contains info about which compiler was used.

Internally we have had to address the situation several times where e.g. tensorflow and related wheels assume quite a lot of things about the native code execution environment (mostly relating to glibc and stdlibc++ version), so if there is some way to tag wheels with this type of more in-depth native code compatibility (e.g. which compiler is used), we could potentially apply that method to 3rdparty wheels as well in the future.

Bazel might have an analogue of a far more general form of compatibility with e.g. [`compatible_with`](https://docs.bazel.build/versions/master/be/common-definitions.html#common.compatible_with) (with application to [bazel toolchains](https://docs.bazel.build/versions/master/toolchains.html#toolchain-resolution)) -- I'm not familiar enough with any of that to say whether it's applicable here. Bazel's compatibility also handles e.g. ensuring "secret" code doesn't get mixed with non-"secret" code, so it's a far more general formulation than I care about right now, but was interesting to look at just now.

Reply:
I've linked the above to `#native` so the content isn't lost. In essence, I'm thinking of ways to propagate the identity (even just the compiler name) of the toolchain we use to the `python_dist()` we build, in a way we can check later in this integration test.

Reply:
That would certainly be preferable. That said, just keep scope in mind. All this to test `NativeTask.get_c_toolchain_variant` and `NativeTask.get_cpp_toolchain_variant` work - which just tests subsystems / options. I think that can be done simply in a unit test (although you do request engine products - not sure how easy that is but I'm pretty sure unit test infra has support for this already). You need not test every bit of program logic through an integration test or unit test or both. Pick your battles based on risk or complexity and this is a tiny skirmish your whacking with a howitzer.

Reply:
So, what's the status of this? The test still log scrapes. Do you want to follow-up with an issue tracking fixing this test and removing the production log.info it requires?

Reply:
Added a link to #6866 to cover my interpretation of this discussion.